sw/device/lib/dif/dif_flash_ctrl.c

```diff
@@ -0,0 +1,1046 @@
+    .index = FLASH_CTRL_FIFO_LVL_PROG_OFFSET,
+};
+
+const static bitfield_field32_t kFifoRdBitfield = {
+    .mask = FLASH_CTRL_FIFO_LVL_RD_MASK, .index = FLASH_CTRL_FIFO_LVL_RD_OFFSET,
+};
+
+/** Helper function to get the offset for an interrupt type. */
+DIF_WARN_UNUSED_RESULT
+static bool get_irq_bit_index(dif_flash_ctrl_irq_t irq_type,
+                              uint8_t *offset_out) {
+  ptrdiff_t offset;
+  switch (irq_type) {
+    case kDifFlashCtrlIrqProgEmpty:
+      offset = FLASH_CTRL_INTR_COMMON_PROG_EMPTY;
+      break;
+    case kDifFlashCtrlIrqProgLevel:
+      offset = FLASH_CTRL_INTR_COMMON_PROG_LVL;
+      break;
+    case kDifFlashCtrlIrqRdFull:
+      offset = FLASH_CTRL_INTR_COMMON_RD_FULL;
+      break;
+    case kDifFlashCtrlIrqRdLevel:
+      offset = FLASH_CTRL_INTR_COMMON_RD_LVL;
+      break;
+    case kDifFlashCtrlIrqOpDone:
+      offset = FLASH_CTRL_INTR_COMMON_OP_DONE;
+      break;
+    case kDifFlashCtrlIrqOpError:
+      offset = FLASH_CTRL_INTR_COMMON_OP_ERROR;
+      break;
+    default:
+      return false;
+  }
+  *offset_out = offset;
+
+  return true;
+}
+
+/** Helper function to check if a transaction is in progress. */
+DIF_WARN_UNUSED_RESULT
+static bool is_busy(const dif_flash_ctrl_t *flash_ctrl) {
+  return mmio_region_get_bit32(flash_ctrl->base_addr,
+                               FLASH_CTRL_CTRL_REGWEN_REG_OFFSET,
+                               FLASH_CTRL_CTRL_REGWEN_EN) == 0;
+}
+
+/** Helper function to set the scramble mode. */
+DIF_WARN_UNUSED_RESULT
+static bool scamble_set(const dif_flash_ctrl_t *flash_ctrl,
+                        dif_flash_ctrl_enable_t scramble_mode) {
+  switch (scramble_mode) {
+    case kDifFlashCtrlEnable:
+      mmio_region_nonatomic_set_bit32(flash_ctrl->base_addr,
+                                      FLASH_CTRL_SCRAMBLE_EN_REG_OFFSET,
+                                      FLASH_CTRL_SCRAMBLE_EN_VAL);
+      break;
+    case kDifFlashCtrlDisable:
+      mmio_region_nonatomic_clear_bit32(flash_ctrl->base_addr,
+                                        FLASH_CTRL_SCRAMBLE_EN_REG_OFFSET,
+                                        FLASH_CTRL_SCRAMBLE_EN_VAL);
+      break;
+    default:
+      return false;
+  }
+
+  return true;
+}
+
+/** Helper function for starting erase flash operation. */
+DIF_WARN_UNUSED_RESULT
+static dif_flash_ctrl_start_result_t start_op(
+    const dif_flash_ctrl_t *flash_ctrl, uint32_t addr, uint32_t op,
+    uint32_t word_count, dif_flash_ctrl_part_sel_t part_sel,
+    dif_flash_ctrl_erase_sel_t erase_sel) {
+  if (is_busy(flash_ctrl)) {
+    return kDifFlashCtrlStartBusy;
+  }
+
+  // Set the address.
+  mmio_region_write32(flash_ctrl->base_addr, FLASH_CTRL_ADDR_REG_OFFSET, addr);
+
+  // Set the operation of the transaction: read, program, or erase.
+  uint32_t control_reg_val = bitfield_field32_write(0, kControlOpBitfield, op);
+
+  // Set the erase mode.
+  switch (erase_sel) {
+    case kDifFlashCtrlEraseSelPage:
+      break;
```

Comment:
Do you mind adding a short comment here similar to the `switch` below to make it clear that this is intended?